File: go/linked_list/141_linked_list_cycle.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func populateListWithCycle(values []int, cyclePos int) *ListNode {
	if len(values) == 0 {
		return nil
	}

	head := ListNode{values[0], nil}
	current := &head
	count := 0

	var cycleNode ListNode
	for _, value := range values[1:] {
		next := ListNode{value, nil}
		current.Next = &next

		if count == cyclePos {
			cycleNode = *current
		}
		count += 1

		current = current.Next

		if count == len(values)-1 {
			current.Next = &cycleNode
		}
	}

	return &head
}

func main() {
	// Test cases

	// result: true
	head := populateListWithCycle([]int{3, 2, 0, -4}, 1)
	fmt.Println(hasCycle(head))

	// result: true
	head = populateListWithCycle([]int{1, 2}, 0)
	fmt.Println(hasCycle(head))

	// result: false
	head = populateListWithCycle([]int{1}, -1)
	fmt.Println(hasCycle(head))
}

func hasCycle(head *ListNode) bool {
	if head == nil {
		return false
	}

	slow := head
	fast := head

	for slow != nil && fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next

		if slow == fast {
			return true
		}
	}

	return false
}

File: ruby/linked_list/141_linked_list_cycle.rb
# Definition for singly-linked list.
class ListNode
  attr_accessor :val, :next
  def initialize(val)
    @val = val
    @next = nil
  end
end

# @param {ListNode} head
# @return {Boolean}
def hasCycle(head)
  slow = head
  fast = head

  while slow && fast && fast.next
    slow = slow.next
    fast = fast.next.next

    return true if slow == fast
  end

  false
end

def populate_list_with_cycle(values = [], cycle_pos)
  return nil if values.empty?

  head = ListNode.new(values[0])
  current = head

  counter = 0
  cycle_node = nil

  values[1..].each do |value|
    nxt = ListNode.new(value)
    current.next = nxt

    cycle_node = current if counter == cycle_pos
    counter += 1

    current = nxt
    current.next = cycle_node if counter == values.size - 1
  end

  head
end

def test
  # Test cases

  # result: true
  puts hasCycle(populate_list_with_cycle([3, 2, 0, -4], 1))

  # result: true
  puts hasCycle(populate_list_with_cycle([1, 2], 0))

  # result: false
  puts hasCycle(populate_list_with_cycle([1], -1))
end

test
